experimental/metal/api.h

```diff
@@ -0,0 +1,107 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// See iree/base/api.h for documentation on the API conventions used.
+
+#ifndef IREE_EXPERIMENTAL_METAL_API_H_
+#define IREE_EXPERIMENTAL_METAL_API_H_
+
+#include "iree/base/api.h"
+#include "iree/hal/api.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+//===----------------------------------------------------------------------===//
+// iree_hal_metal_device_params_t
+//===----------------------------------------------------------------------===//
+
+typedef enum iree_hal_metal_command_dispatch_type_e {
+  // Dispatching commands in command buffer in parallel.
+  IREE_HAL_METAL_COMMAND_DISPATCH_TYPE_CONCURRENT = 0,
+  // Dispatching commands in command buffer sequentially.
+  IREE_HAL_METAL_COMMAND_DISPATCH_TYPE_SERIAL = 1,
+} iree_hal_metal_command_dispatch_type_t;
+
+typedef enum iree_hal_metal_command_buffer_resource_reference_mode_e {
+  // Letting the command buffer to not maintain strong references to used
+  // resources.
+  IREE_HAL_METAL_COMMAND_BUFFER_RESOURCE_REFERENCE_MODE_UNRETAINED = 0,
+  // Letting the command buffer to maintain strong references to used resources.
+  IREE_HAL_METAL_COMMAND_BUFFER_RESOURCE_REFERENCE_MODE_RETAINED = 1,
+} iree_hal_metal_command_buffer_resource_reference_mode_t;
+
+typedef enum iree_hal_metal_resource_hazard_tracking_mode_e {
+  // Letting the app to prevent hazards when modifying this object's contents.
+  IREE_HAL_METAL_RESOURCE_HAZARD_TRACKING_MODE_UNTRACKED = 0,
+  // Letting the Metal runtime to prevent hazards when modifying this object's
+  // contents.
+  IREE_HAL_METAL_RESOURCE_HAZARD_TRACKING_MODE_TRACKED = 1,
+} iree_hal_metal_resource_hazard_tracking_mode_t;
```

Comment:
I'm not sure I follow here. Is "Metal runtime" something we're using, or is it the IREE Metal HAL driver itself?

```c
  // Resource hazard tracking mode.
  // IREE is following explicit GPU API model and tracks resource dependency by
  // itself. So normally we don't need to let Metal runtime to track resource
  // usages and prevent hazards, which incurs runtime overhead. But it can be
  // helpful for debugging purposes.
  iree_hal_metal_resource_hazard_tracking_mode_t resource_hazard_tracking_mode;
```

```suggestion
typedef enum iree_hal_metal_resource_hazard_tracking_mode_e {
  // Do not track resource hazards. Hosting applications are responsible
  // for ensuring that resources are not modified while in use.
  IREE_HAL_METAL_RESOURCE_HAZARD_TRACKING_MODE_UNTRACKED = 0,
  // Track resource hazards within the Metal runtime.
  IREE_HAL_METAL_RESOURCE_HAZARD_TRACKING_MODE_TRACKED = 1,
} iree_hal_metal_resource_hazard_tracking_mode_t;
```

Reply:
"Metal runtime" means the underlying Metal runtime in the system; not IREE Metal driver. I added "system" to prefix to be clear.